### ISSUE TITLE
Is `crossentropy()` correct?

### ISSUE THREAD

**Issue by SimonEnsemble** [FluxML/Flux.jl#139]
As implemented:
```
function crossentropy(ŷ::AbstractVecOrMat, y::AbstractVecOrMat; weight = 1)
  return -sum(y .* log_fast.(ŷ) .* weight) / size(y, 2)
end
```

I took a look at the code b/c I was getting `Inf`; I think in practice ppl add a small number to the argument of the `log` to avoid this.

But I think [binary] cross entropy should be:
```
const eps = 1e-7

function crossentropy(ŷ::AbstractVecOrMat, y::AbstractVecOrMat)
    return -sum(y .* log.(ŷ + eps)) - sum((1.0 - y) .* log.(1.0 - ŷ + eps))
end
```

See [here](https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_error_function_and_logistic_regression).

**Comment by CarloLucibello** [comment]
Crossentropy and binary crossentropy are two slightly different things. Crossentropy is meant to by used on the output of a softmax layer which encodes a probability distributions over K classes. The binary crossentropy you wrote should be applied on the output of a logistic or a  tanh layer, encoding a factorized probability over 2^K binary configurations. So, both methods you wrote express cross entropies, but they are meant to be applied to different probabilistic spaces. 

In my experience (with Knet, but I guess with Flux is the same) I never encountered inf/nan in the crossentropy, unless initializing with weights too big or setting a learning rate too high. The binary crossentropy instead is more fragile, and yes adding an epsilon fixes the problem.

**Comment by SimonEnsemble** [comment]
Thanks @CarloLucibello. So for two class discrimination, I should use the one-hot encoding such as 
```
y = [0 0 1; 1 1 0]
```
(for 3 data points) if I want to use the `crossentropy` loss function.

Otherwise, if I encode the target values as `y = [1 1 0]`, I will just write my own binary cross entropy loss function as above.

**Comment by MikeInnes** [comment]
Correct. Flux could house a binary crossentropy function as well if you want to contribute it.

**Comment by MikeInnes** [comment]
Closing this in favour of #143
